storage/buffer.go

```diff
@@ -378,7 +381,11 @@ func (it *SampleRingIterator) AtFloatHistogram(fh *histogram.FloatHistogram) (in
 	if it.fh == nil {
 		return it.t, it.h.ToFloat(fh)
 	}
-	return it.t, it.fh
+	if fh != nil {
+		fh.CopyFrom(*it.fh)
+		return it.t, fh
+	}
+	return it.t, it.fh.Copy()
```

Comment:
This copy is new, is it there for consistency or was there a bug?

Reply:
This copy is not necessary here, but I thought it is safer to do it than not to. When we create a new histogram slice with `getHPointSlice`, the items in it will be `nil` pointers. If we pass in `nil` as the argument, we will get a reference to a sample in the buffer, and it will be easy to overwrite it by mistake if we use the sample as an argument to `floatHistogramIterator.AtFloatHistogram`. 

The alternative would be to never use `nil` as the argument and allocate all histogram objects at the call site (in `matrixIterSlice`) so that we trigger the `CopyFrom` branch. I would be fine with either solution, this was just the simplest one to implement.

Reply:
I have the feeling that this takes away some optimization possibilities to delay the copy (or avoid copy if the caller uses the data as read only) and is overly defensive. On the other hand it's for sure safer. So I guess as long as performance is not hit, it's fine.

Reply:
I agree, it's definitely suboptimal and something we can look into next. However, I did some local testing with the new branch and the copy here was needed unless I passed in an object that could be filled. Right now histograms tests are not comprehensive enough to catch regressions of this sort. 

I think the better solution is for `getHPointSlice` to return a slice of value objects rather than pointers, so that when we get a slice it will always have non-nil values.

Reply:
ack